[DOCS] Add clarifications to mapping type

ChrisChinchilla opened the pull request:
### Description

Closes https://github.com/ethereum/solidity/issues/6303 by adding some description around the code example, and adding a link to the type when mentioned in other places.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -30,7 +30,13 @@ You can mark state variables of mapping type as ``public`` and Solidity creates
 parameter for the getter. If ``_ValueType`` is a value type or a struct,
 the getter returns ``_ValueType``.
 If ``_ValueType`` is an array or a mapping, the getter has one parameter for
```
This doesn't really make sense. Which `_KeyType`? The getter has additional parameters as needed to access the final element in the mapping / array structure.

ChrisChinchilla replied:
@chriseth it refers to the paragraph a couple above that uses the same analogies, but I'll make the connection between the two paragraphs clearer.

Marenz replied:
well, the keytype is mentioned in line 7. But for clarification we could write "the mappings keytype" / "the keytype of the mapping"
On that note, maybe a cod example is useful here, too

ChrisChinchilla replied:
There is a code example a paragraph later, might be better to view the overall changes to see everything in context and let me know if you think anything is still needed.

chriseth replied:
Ah sorry, now I see it!

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -30,7 +30,13 @@ You can mark state variables of mapping type as ``public`` and Solidity creates
+
+In the example below, the ``MappingExample`` contract defines a public ``balances``
+mapping, with the key type an ``address``, and a value type a ``uint``, mapping
+an integer value to an Ethereum address. As ``uint`` is a value type, the getter
```
Mapping an address to an unsigned integer.

ChrisChinchilla commented:
@chriseth I understand your last comment, and I understand the principles, but as this was the content that was already there, I'm struggling slightly to explain it better.

The pseudo syntax refers to the earlier paragraph, and I am hoping that the example (and explanation) help explain it…

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -4,7 +4,7 @@
 =============
 
-You declare mapping types with the syntax ``mapping(_KeyType => _ValueType)``.
+You declare mapping types with the syntax ``mapping(_KeyType => _ValueType) _VariableModifiers _VariableName``.
```
This is how you declare a variable of mapping type, not how you declare a mapping type.

Maybe `declare` is also the wrong word, `use` may be better.

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -26,11 +26,16 @@ They cannot be used as parameters or return parameters
-:ref:`getter <visibility-and-getters>` for you. The ``_KeyType`` becomes a
-parameter for the getter. If ``_ValueType`` is a value type or a struct,
-the getter returns ``_ValueType``.
+:ref:`getter <visibility-and-getters>` for you. The ``_KeyType`` is the parameter for the getter.
```
Please keep "becomes a", because it may not be the only parameter.

ChrisChinchilla commented:
Changes made

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -4,7 +4,8 @@
 
-You declare mapping types with the syntax ``mapping(_KeyType => _ValueType)``.
+Mapping types use the syntax ``mapping(_KeyType => _ValueType)`` and variables
+are declared as a mapping type using the syntax ``mapping (_KeyType => _ValueType) _VariableModifiers _VariableName``.
```
It's probably not too useful to keep this.

chriseth replied:
I still do not think it is useful to keep this.

ChrisChinchilla replied:
Which bit? It was here already, and from the conversations I've had around, people are really confused by the mapping syntax, so I'd love to try and be as explicit and clear as possible. Any suggestions to keep that clarity but express it differently?

chriseth replied:
I think it would be better to give an example how to declare a specific variable instead of just adding random placeholders whose meaning is not explained.

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -26,11 +27,16 @@ They cannot be used as parameters or return parameters
-each ``_KeyType``, recursively. For example with a mapping:
+each ``_KeyType``, recursively.
+
+In the example below, the ``MappingExample`` contract defines a public ``balances``
```
Please also add a mapping of a mapping to the example.

ChrisChinchilla commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
+point to remember are that the mapping inside the mapping is not a return variable,
+but the variable inside it is.
+
+::
```
@chriseth Myself and @ekpyron had a chat about an example for a mapping with a mapping and struggled to think of many great examples. What do you think of this one? Then I'll polish the text.

chriseth replied:
looks good!

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
 
+The example below uses a mapping type inside another mapping type. The important
+point to remember are that the mapping inside the mapping is not a return variable,
+but the variable inside it is.
```
"value" instead of 'variable' for both?

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
+    contract MappingUser {
+        function f() public returns (bool) {
+            MappingExample m = new MappingExample();
+            m.transfer(address(this),address(this));
```
```suggestion
            m.transfer(address(this), address(this));
```

chriseth replied:
maybe at least transfer from this to msg.sender? Or provide an address as a parameter of `f()`? Then you do not need the type conversions.

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
+        function f() public returns (bool) {
+            MappingExample m = new MappingExample();
+            m.transfer(address(this),address(this));
+            return m.validBalances(address(this),address(this));
```
```suggestion
            return m.validBalances(address(this), address(this));
```

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
+    pragma solidity >=0.4.0 <0.7.0;
+
+    contract MappingExample {
+        mapping(address => mapping(address => bool)) public validBalances;
```
Doesn't erc20 use something like `allowances`?

ChrisChinchilla replied:
It does, but I'm not sure what you mean? This isn't really an ERC20 implementation, and the data structure here is different too.

chriseth replied:
No, it's exactly the same - and I think it would be a good example: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L36

ChrisChinchilla commented:
@chriseth Made changes, i think I corrected another code error too.

chriseth commented:
Tests are failing.

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -26,11 +27,16 @@ They cannot be used as parameters or return parameters
+each ``_KeyType``, recursively.
+
+In the example below, the ``MappingExample`` contract defines a public ``balances``
+mapping, with the key type an ``address``, and a value type a ``uint``, mapping
```
Isn't it mapping the address to the uint?

ChrisChinchilla replied:
I think that is what I was trying to say, but I'll make it clearer.

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
+
+    pragma solidity >=0.4.0 <0.7.0;
+
+    contract MappingExample {
```
Do you think a full credit / allowance example would be too long?

ChrisChinchilla replied:
@chriseth Possibly, but how much do you think the full example relates to mappings and how they work? Or maybe a reduced example of what you suggest that highlights those aspects? I assume based on the Zeppelin link from above?

chriseth replied:
I think you can use the following functions and also slim them down a little:
 - approve
 - transferFrom 
 - allowances

ekpyron commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,34 @@ each ``_KeyType``, recursively. For example with a mapping:
     }
 
+The example below uses a mapping type inside another mapping type. The important
+point to remember are that the mapping inside the mapping is not the return value,
```
That should be ``The important point to remember is`` not ``are``... but also: "the return value" of what? I'm not sure I understand the purpose of this sentence :-).

ChrisChinchilla replied:
Corrected that. The point is to clarify that the returned value isn't the mapping inside the mapping, but the value inside the nested mapping. That's what I understood from you anyway?

ekpyron replied:
When you're talking about "the returned value", you have to say what you are talking about that returns something - the index access of the outer mapping? The index access of the outer mapping in fact *does* return "the mapping inside the mapping" - only the second index access yields the value contained in the inner mapping...

```
contract C {
    mapping(address => mapping(address => bool)) public allowances;
    function f(address X, address Y) public view {
        mapping(address => bool) storage allowancesForX = allowances[X]; // first index access is a mapping
        bool allowanceForXandY = allowances[X][Y]; // two index accesses give the contained value
        bool allowanceForXandY2 = allowancesForX[Y]; // which is the same as accessing the contained mapping at Y
        assert(allowanceForXandY == allowanceForXandY2);
    }
}

```

ChrisChinchilla commented:
OK @chriseth done!

I had to cut a fair bit out, so I hope nothing got lost in the process. @ekpyron this also cut most of the copy that we were discussing, possibly it even needs more explanation now!

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,54 @@ each ``_KeyType``, recursively. For example with a mapping:
 
+The example below is a simplified version of an `ERC20 token <https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol>`_.
+``_allowances`` is an example of a mapping type inside another mapping type.
+The example below uses ``_allowances`` to record the amount of a transaction and the two addresses it was between.
```
```suggestion
The example below uses ``_allowances`` to record the amount someone else is allowed to withdraw from your account.
```

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -52,6 +58,54 @@ each ``_KeyType``, recursively. For example with a mapping:
+
+        function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
+            _transfer(sender, recipient, amount);
+            _approve(sender, msg.sender, amount);
```
Why do you call approve here? It should rather check that the allowance is large enough.

ChrisChinchilla replied:
@chriseth This came from the Zeppelin code…

chriseth replied:
But it does something else now.

erak commented:
Needs to rebased since the failing test was fixed with https://github.com/ethereum/solidity/pull/7419.

erak reviewed (changes requested):
Needs to rebased since the failing test was fixed with #7419.

ChrisChinchilla commented:
@erak Rebased, squashed

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -27,11 +28,16 @@ They cannot be used as parameters or return parameters
+In the example below, the ``MappingExample`` contract defines a public ``balances``
+mapping, with the key type an ``address``, and a value type a ``uint``, mapping
+an Ethereum address to an unsigned integer value. As ``uint`` is a value type, the getter
+returns a value that matches the type, which you can see in the ``MappingUser``
```
"of that type" (how does a value match a type?)

chriseth commented on docs/types/mapping-types.rst:
```diff
@@ -53,6 +59,54 @@ each ``_KeyType``, recursively. For example with a mapping:
+            emit Transfer(sender, recipient, amount);
+        }
+
+        function _approve(address owner, address spender, uint256 amount) internal {
```
I think it is better to combine this function with `approve` and handle the approval reduction in transferFrom manually.

ChrisChinchilla replied:
I think this is what you're asking for…

Commit:
Explain mappings example, and add link to further details

Updates from review

Changes from review

Nested example

Update docs/types/mapping-types.rst

Co-Authored-By: chriseth <chris@ethereum.org>
Changes from review

Bring example inline with ERC20

Clarify what maps where

Use OZ contract example

Update docs/types/mapping-types.rst

Co-Authored-By: chriseth <chris@ethereum.org>

update code example

chriseth commented:
Rebased.

chriseth reviewed (approved):
Still not sure about some details, but better to merge this.
